test/common/testlib.py

```diff
@@ -743,27 +761,32 @@ def setUp(self):
         if opts.address and self.provision is not None:
             raise unittest.SkipTest("Cannot provision multiple machines if a specific machine address is specified")
 
-        self.machine = None
         self.browser = None
         self.machines = {}
         provision = self.provision or {'machine1': {}}
 
-        # First create all machines, wait for them later
-        for key in sorted(provision.keys()):
-            options = provision[key].copy()
-            if 'address' in options:
-                del options['address']
-            if 'dns' in options:
-                del options['dns']
-            if 'dhcp' in options:
-                del options['dhcp']
-            machine = self.new_machine(**options)
-            self.machines[key] = machine
-            if not self.machine:
-                self.machine = machine
-            if opts.trace:
-                print("Starting {0} {1}".format(key, machine.label))
-            machine.start()
+        test_method = getattr(self.__class__, self._testMethodName)
+        if getattr(test_method, "_testlib__non_destructive", False) and not opts.address:
+            self.machines['machine1'] = MachineCase.get_global_machine()
+            self.machine = MachineCase.get_global_machine()
+        else:
+            self.machine = None
+            # First create all machines, wait for them later
+            for key in sorted(provision.keys()):
+                options = provision[key].copy()
```

Comment:
I hesitate to suggest a dictionary comprehension here since it's less readable, but it somehow feels "cleaner" than copying and deleting from the copy:

    filter = ['address', 'dns', 'dhcp']
    for name in sorted(provision.keys()):
        options = {key:provision[name][key] for key in provision[name] if key not in filter}

Definitely a "matter of taste" type of call...

Reply:
so this was pre-existing code and I prefer to not change any chunks that were working fine. But if you feel strongly I'll change it.

Reply:
I don't :)